### ISSUE TITLE
allow variable height front pages

### ISSUE THREAD

**Issue by mockturtl** [fluttercommunity/backdrop#70]
**Is your feature request related to a problem? Please describe.**

My first front page is a large form.  It displays well.

![Screenshot from 2021-01-10 17-34-12](https://user-images.githubusercontent.com/705804/104137374-23f79900-536a-11eb-85e4-4216077b2ab1.png)

My second front page is mostly empty:

![Screenshot from 2021-01-10 17-33-08](https://user-images.githubusercontent.com/705804/104137387-38d42c80-536a-11eb-82e4-34e09ba05434.png)

I'd like to provide a target height for this page, so it expands to a more appropriate size for its content.

![Screenshot from 2021-01-10 17-32-50](https://user-images.githubusercontent.com/705804/104137403-51dcdd80-536a-11eb-9b51-f2185f3eae19.png)

**Describe the solution you'd like**

I can get my desired behavior by adjusting the `top` param of the `end` rectangle in [`_getPanelAnimation`](https://github.com/fluttercommunity/backdrop/blob/master/lib/scaffold.dart#L458).

Could we expose a variable for this?

**Describe alternatives you've considered**

I haven't explored how this will interact with the other `BackdropScaffold` constructor args.

Maybe `PreferredSizeWidget` is a useful way to convey the behavior?

**Additional context**

The Material [spec](https://material.io/components/backdrop#specs) doesn't address whether front pages should have a consistent height.

**Comment by daadu** [comment]
@mockturtl Have you tried `stickyFrontLayer` and `headerHeight` arguments of `BackdropScaffold`, to get desired UI?

**Comment by mockturtl** [comment]
- Playing with values in `example/lib/navigation.dart`, my understanding of the [`stickyFrontLayer`](https://pub.dev/documentation/backdrop/latest/scaffold/BackdropScaffold/stickyFrontLayer.html) behavior is "When closed, the front layer will occupy as much height as possible without occluding the back layer.  If the back layer occupies the entire screen height[1], this parameter has no effect."

[1] This is my use case: my back layer is a large menu.  Apologies for misrepresenting that in the screenshot above.

- [`headerHeight`](https://pub.dev/documentation/backdrop/latest/scaffold/BackdropScaffold/headerHeight.html) seems to control the height of the front layer in its **inactive / closed** state (this contradicts my reading of the docs[2]).  Additionally, if `stickyFrontLayer == true` and `headerHeight` is less than the available space, `headerHeight` is ignored.

[2] It might be desirable to document the control flow inside the `_headerHeight` getter and `_getPanelAnimation`, if it's not too cumbersome to explain.

----

From what I can tell, the existing constructor args for `BackdropScaffold` don't allow me to specify "front page A opens to height X, front page B opens to height Y."

@willlarche Does Material have an opinion on a Backdrop with variable-height front pages?  I'll double check [Crane](https://material.io/design/material-studies/crane.html#product-architecture).

**Comment by daadu** [comment]
@mockturtl Thanks for pointing out problem in `headerHeight` docs, will correct it.

Regarding your requirement, as far as I understand(from checking Crane example), you want to have "extra" space under appbar, that is visible even when the `backLayer` is concealed.

A solution for that would be to set `bottom`  of `BackdropAppBar`. Check the following code (modified example/main.dart):

```dart
MaterialApp(
      title: 'Backdrop Demo',
      home: BackdropScaffold(
        appBar: BackdropAppBar(
          title: Text("Backdrop Example"),
          actions: <Widget>[
            BackdropToggleButton(
              icon: AnimatedIcons.list_view,
            )
          ],
          bottom: PreferredSize(
            child: Column(
              children: [
                Text("Menu 1"),
                Text("Menu 2"),
              ],
            ),
            preferredSize: Size.fromHeight(100),
          ),
        ),
        backLayer: Center(
          child: Text("Back Layer"),
        ),
        subHeader: BackdropSubHeader(
          title: Text("Sub Header"),
        ),
        frontLayer: Center(
          child: Text("Front Layer"),
        ),
      ),
    );
```
Screenshot for it:
![untitled](https://user-images.githubusercontent.com/4963236/104280853-908da780-54d2-11eb-885d-f725b3c72456.gif)


As far as I understand, the "extra" (always visible stuff) is part of AppBar and not backLayer (which is revealed and concealed)

**Comment by willlarche** [comment]
@mockturtl Hi! So wonderful to see this implementation. Much thanks to everyone working on expanding Material coverage in the community.

We have no issue with that! Go for it!

**Comment by mockturtl** [comment]
> Regarding your requirement, as far as I understand(from checking Crane example), you want to have "extra" space under appbar, that is visible even when the backLayer is concealed

Not really.  I want the front layer's height to be configurable for each page.  I'm not sure how to explain differently.

# Back layer and appbar look good

![Screenshot from 2021-01-12 15-36-27](https://user-images.githubusercontent.com/705804/104369844-0e11e180-54ec-11eb-9377-dce823a69b4f.png)

# Front layer: Current behavior

![Screenshot from 2021-01-12 15-36-32](https://user-images.githubusercontent.com/705804/104369876-179b4980-54ec-11eb-929e-8014266a5371.png)

![Screenshot from 2021-01-12 15-31-34](https://user-images.githubusercontent.com/705804/104369928-2681fc00-54ec-11eb-8f07-ba800381b2b7.png)

# Front layer: Desired behavior

![Screenshot from 2021-01-12 15-47-03](https://user-images.githubusercontent.com/705804/104370862-7a411500-54ed-11eb-9e99-9c1b347b205a.png)

**Comment by daadu** [comment]
@mockturtl Ok, got your point. Go ahead with the PR! BTW, what should we name the param? Any idea? 

Few of my suggestions:
  - `double frontLayerStartFrom: 0`
  - `double frontLayerTopMargin: 0`
  - ??

Also while filing PR can you correct the docs for `headerHeight` - instead of using "closed" and "open" term (earlier we used them), use phrases like "when `backLayer` is concealed/revealed", that should avoid the confusion with the usage of `headerHeight.

**Comment by daadu** [comment]
This has landed with [v0.5.0 on pub.dev](https://pub.dev/packages/backdrop/versions/0.5.0).
